lib/runCall.js

```diff
@@ -59,111 +61,287 @@ module.exports = function (opts, cb) {
+
+    updateAccounts(modifiedAccounts, () => {
+      var runState = {
+        blockchain: self.blockchain,
+        stateManager: stateManager,
+        storageReader: storageReader,
+        returnValue: returnData,
+        stopped: false,
+        vmError: false,
+        programCounter: opts.pc | 0,
+        opCode: undefined,
+        opName: undefined,
+        gasLeft: gasRemaining,
+        gasLimit: gasLimit,
+        gasPrice: gasPrice,
+        memory: new Memory(),
+        memoryWordCount: new BN(0),
+        stack: new Stack(),
+        lastReturned: [],
+        logs: logs,
+        validJumps: [],
+        gasRefund: gasRefund,
+        highestMemCost: new BN(0),
+        depth: depth,
+        selfdestruct: selfdestruct,
+        block: block,
+        callValue: opts.value || new BN(0),
+        address: opts.address || ethUtil.zeros(32),
+        caller: caller,
+        origin: opts.origin || opts.caller || ethUtil.zeros(32),
+        callData: opts.data || Buffer.from([0]),
+        code: code,
+        static: opts.static || false
+      }
+
+      vmResults = {
+        runState: runState,
+        selfdestruct: runState.selfdestruct,
+        gasRefund: runState.gasRefund,
+        exception: error ? 0 : 1,
+        exceptionError: error,
+        logs: runState.logs,
+        gas: runState.gasLeft,
+        'return': runState.returnValue ? runState.returnValue : Buffer.alloc(0)
       }
-      try {
-        query = msg_pb.VMQuery.deserializeBinary(mdata)
-        if (CALLRESULT === query.getQueryCase()) {
-          console.log('CALLRESULT', query)
-        }
-      } catch (err) {
-        console.log(err)
+
+      results = {
+        gasUsed: gasProvided.sub(gasRemaining),
+        createdAddress: createdAddress,
+        vm: vmResults
       }
+      _callback()
+    })
+  }
+
+  function updateAccounts (accountList, _callback) {
+    async.series(
+      accountList.map(account => function (_itemCallback) {
+        updateAccount(account, _itemCallback)
+      })
+      , _callback
+    )
+  }
+
+  function commitState () {
+    stateManager.commit(function () {
+      console.log('commit done')
+      cb(null, results)
+    })
+  }
+
+  function updateAccount (pbAccount, _done) {
+    var address = Buffer(pbAccount.getAddress())
+    var nonce = Buffer(pbAccount.getNonce())
+    var newBalance = Buffer(pbAccount.getBalance())
+    var code = Buffer(pbAccount.getCode())
+    var storageUpdateList = pbAccount.getStorageupdatesList()
+
+    async.series(
+      [function (_callback) {
+        stateManager.getAccount(address, function (err, stateAccount) {
+          if (err) {
+            console.log(err)
+          }
+          stateAccount.nonce = Buffer(nonce)
+          stateAccount.balance = Buffer(newBalance)
+          stateManager.putAccount(address, stateAccount, _callback)
+        })
+      },
+        function (_callback) {
+          stateManager.putContractCode(address, code, _callback)
+        },
```

Comment:
We should be checking if `code == ""`, and if so not update the contract code.

Reply:
Added a new condition to check the `code` length.